include/ghex/transport_layer/context.hpp

```diff
@@ -139,39 +31,34 @@ namespace gridtools {
                 using tag                    = TransportTag;
                 using transport_context_type = transport_context<tag,ThreadPrimitives>;
                 using communicator_type      = typename transport_context_type::communicator_type;
-                using parallel_context_type  = parallel_context<ThreadPrimitives>;
-                using thread_primitives_type = typename parallel_context_type::thread_primitives_type;
-                using thread_token           = typename parallel_context_type::thread_token;
+                using thread_primitives_type = ThreadPrimitives;
+                using thread_token           = typename thread_primitives_type::token;
 
             private:
-                parallel_context_type  m_parallel_context;
+                thread_primitives_type m_thread_primitives;
                 transport_context_type m_transport_context;
 
             public:
                 template<typename...Args>
-                context(int num_threads, MPI_Comm comm, Args&&... args)
-                : m_parallel_context{num_threads, comm, std::forward<Args>(args)...}
-                , m_transport_context{m_parallel_context, comm, std::forward<Args>(args)...}
+                context(int num_threads, Args&&... args)
+                    : m_thread_primitives(num_threads)
+                    , m_transport_context{m_thread_primitives, std::forward<Args>(args)...}
                 {}
 
                 context(const context&) = delete;
                 context(context&&) = delete;
 
             public:
 
-                const mpi_world& world() const noexcept 
+                /*const */thread_primitives_type& thread_primitives() /*const*/ noexcept
                 {
-                    return m_parallel_context.world(); 
-                }
-                
-                /*const */thread_primitives_type& thread_primitives() /*const*/ noexcept 
-                {
-                    return m_parallel_context.thread_primitives(); 
+                    return m_thread_primitives;
                 }
 
                 mpi::setup_communicator get_setup_communicator()
                 {
-                    return mpi::setup_communicator(m_parallel_context.world());
+                    // TODO: what about this?
```

Comment:
@boeschf This is hanging: since we do not have the mpi comm here anymore, the setup communicator should be moved elsewhere. Do you need it here?

Reply:
Yes I do need it, this should maybe ask the transport context for the mpi-comm?

Reply:
@boeschf the transport context doesn't have the MPI comm :) at least not the UCX context

Reply:
there needs to be a mapping to an mpi comm at some point somewhere - that's a requirement we have for the high level API (to be able to figure out halos, to be able to do collectives, etc.). We need to rethink this initialization business